### PR TITLE
bpo-27334: roll back transaction if sqlite3 context manager fails to commit

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue27334

### CODE REVIEW

**Comment by erlend-aasland** [comment]
@berkerpeksag, a couple of comments accompanying the change:

1. Since the test calls for another process (I can't reproduce it using threads), I added a new test class `MultiprocessTests` for this. If there's a better way, shout out :)
2. AFAICS, the test should be able to clean up both the test database file and the child process, regardless of where it may fail. The first `input()` in the child process should take care of any race conditions.
3. I swapped `method_name` with a flag for performance reasons; it's easier to test for a flag when `result == NULL`
4. Because of 3., it was easier to just call the commit/rollback functions directly, iso. going through a `PyObject_Call`; hope that's ok. As a side effect, `__exit__` should now be a tiny bit faster.

**Comment by erlend-aasland** [comment]
(cc. @pablogsal)

**Comment by lciti** [comment]
> I've taken the liberty to create a PR based on your patch, Luca. Berker's comments have been addressed in the PR.

Thanks for creating the PR. Sorry for not doing it myself, I almost forgot about the whole thing. It's good that a fix is eventually making its way into the code.

**Comment by erlend-aasland** [comment]
> Thanks for creating the PR. Sorry for not doing it myself, I almost forgot about the whole thing. It's good that a fix is eventually making its way into the code.

Thanks for reporting, providing a reproducer, and proposing a fix! :) I've credited you in the NEWS entry and in the commit message.

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [inline Modules/_sqlite/connection.c]: Alternative approach where we return `False` if rollback succeeds, iso. just returning `NULL`. I don't think this really matters, the exception is raised either way.

```suggestion
        if (commit) {
            result = pysqlite_connection_rollback_impl(self);
        }
        if (result == NULL) {
            return NULL;
        }
```

File: Lib/sqlite3/test/dbapi.py
Comment (pablogsal) [inline Lib/sqlite3/test/dbapi.py]: Maybe use TIMEOUT here?

File: Lib/sqlite3/test/dbapi.py
Comment (pablogsal) [inline Lib/sqlite3/test/dbapi.py]: Why this small TIMEOUT value? (Notice the test suite uses some customizable timeout values that yuou may want to reuse here)

File: Modules/_sqlite/connection.c
Comment (pablogsal) [reply Modules/_sqlite/connection.c]: You cannot return a valid python object with an exception being set. Maybe i am misunderstanding what you mention with "return False"

File: Lib/sqlite3/test/dbapi.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: > Why this small TIMEOUT value?

The default connection timeout is very long. The test depends on SQLite timing out when it tries to commit. Using a small timeout value makes it possible to execute the test very fast, thus not slowing down the CI.

> (Notice the test suite uses some customizable timeout values that yuou may want to reuse here)

Oh, I did not know; I'll check it out. Thanks :)

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: This `__exit__` method defaults to returning `Py_RETURN_FALSE` a few lines below. So `result != NULL` => `Py_RETURN_FALSE`, and `result == NULL` => `return NULL`.

Currently this PR returns `NULL` if commit fails but rollback succeeds. Alternatively we can use `Py_RETURN_FALSE` if commit fails but rollback succeeds. I'm not sure it matters.

Comment (pablogsal) [reply Modules/_sqlite/connection.c]: > Currently this PR returns `NULL` if commit fails but rollback succeeds. Alternatively we can use `Py_RETURN_FALSE` if commit fails but rollback succeeds. I'm not sure it matters.

As long as we are always returning NULL with an exception set, then we are fine.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: `pysqlite_connection_rollback_impl` can call `PyErr_Clear`, so I added a fix for that. If rollback also fails, we overwrite the current exception with `PyErr_SetString`.

File: Modules/_sqlite/connection.c
Comment (pablogsal) [inline Modules/_sqlite/connection.c]: Can you delete this variable?

File: Lib/sqlite3/test/dbapi.py
Comment (pablogsal) [inline Lib/sqlite3/test/dbapi.py]: Just checking: you want to keep this or you missed changing this to `SHORT_TIMEOUT`?

Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: I want to keep it. `SHORT_TIMEOUT` is way too long (default 30.0). I can use if I divide by 100. Perhaps I should rename it to `SQLITE_TIMEOUT`, or `CONNECTION_TIMEOUT` then.

Comment (pablogsal) [reply Lib/sqlite3/test/dbapi.py]: Yes, I would prefer to have it in a way we can configure it easily and maybe reuse it.

Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: (Um, there was some context missing from my previous post. Added now.)

Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: > Yes, I would prefer to have it in a way we can configure it easily and maybe reuse it.

I agree. Fixed in f67591cc421ef1fcac915bda39edd492c1b38d87

Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: PTAL :)

File: Modules/_sqlite/connection.c
Comment (vstinner) [inline Modules/_sqlite/connection.c]: Is it possible to chain commit and rollback exceptions?

File: Misc/NEWS.d/next/Library/2021-05-18-00-17-21.bpo-27334.32EJZi.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2021-05-18-00-17-21.bpo-27334.32EJZi.rst]: This description sounds too low-level to me. Maybe just that a rollback is run if the commit failed?

Comment (vstinner) [reply Misc/NEWS.d/next/Library/2021-05-18-00-17-21.bpo-27334.32EJZi.rst]: You can keep the sentence about the "database lock" in a parenthesis ;-)

Comment (erlend-aasland) [reply Misc/NEWS.d/next/Library/2021-05-18-00-17-21.bpo-27334.32EJZi.rst]: Yes, this is too low level. I'll rewrite asap :)

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: I don't know! 😀 Let me try and find out.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Hm, PEP 3134 seems to be obsolete, and PEP 490 was rejected, but it seems I can use ~`PyErr_NormalizeException`~ `PyException_SetTraceback` for this? :)

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: Aha, I found `PyException_SetCause` and `PyException_SetContext`. PEP 3134 is apparently not dead.

**Comment by vstinner** [comment]
You can try `_PyErr_ChainExceptions()` to chain exceptions.

Example:

```
static PyObject *
_io_FileIO_close_impl(fileio *self)
/*[clinic end generated code: output=7737a319ef3bad0b input=f35231760d54a522]*/
{
    PyObject *res;
    PyObject *exc, *val, *tb;
    int rc;
    _Py_IDENTIFIER(close);
    res = _PyObject_CallMethodIdOneArg((PyObject*)&PyRawIOBase_Type,
                                       &PyId_close, (PyObject *)self);
    if (!self->closefd) {
        self->fd = -1;
        return res;
    }
    if (res == NULL)
        PyErr_Fetch(&exc, &val, &tb);
    if (self->finalizing) {
        PyObject *r = fileio_dealloc_warn(self, (PyObject *) self);
        if (r)
            Py_DECREF(r);
        else
            PyErr_Clear();
    }
    rc = internal_close(self);
    if (res == NULL)
        _PyErr_ChainExceptions(exc, val, tb);
    if (rc < 0)
        Py_CLEAR(res);
    return res;
}
```

**Comment by erlend-aasland** [comment]
> You can try `_PyErr_ChainExceptions()` to chain exceptions.

Thanks!

File: Modules/_sqlite/connection.c
Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: I tried implementing this in 98d3adaf5bc02ce0db2fd93246af9d5d4139a508. PTAL.

Comment (erlend-aasland) [reply Modules/_sqlite/connection.c]: We should probably chain exceptions in the first part of `__exit__` as well:

1. `__exit__` is called with an exception
2. rollback fails => chain exceptions

File: Lib/sqlite3/test/dbapi.py
Comment (vstinner) [inline Lib/sqlite3/test/dbapi.py]: What is the purpose of this timeout?

File: Lib/sqlite3/test/dbapi.py
Comment (vstinner) [inline Lib/sqlite3/test/dbapi.py]: Do you expect the DB to be closed implicitly?

File: Lib/sqlite3/test/dbapi.py
Comment (vstinner) [inline Lib/sqlite3/test/dbapi.py]: Hum, I don't know if `\n` works on Windows and macOS. Maybe use print(file=...)?

File: Lib/sqlite3/test/dbapi.py
Comment (vstinner) [inline Lib/sqlite3/test/dbapi.py:1059]: I would prefer to get an exception if the child takes longer than 30 seconds to quit. So I suggest to replace except with "except: (...) raise".

File: Lib/sqlite3/test/dbapi.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: It's the SQLite [busy timeout](https://sqlite.org/c3ref/busy_timeout.html). In `sqlite3`, it defaults to 5 seconds. If we want this test to execute fast, we need a short timeout, since the test depends on SQL functions that times out :)

File: Lib/sqlite3/test/dbapi.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: Good catch; I'll close it explicitly. Thanks.

File: Lib/sqlite3/test/dbapi.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py]: Good point! Thanks.

File: Lib/sqlite3/test/dbapi.py
Comment (erlend-aasland) [reply Lib/sqlite3/test/dbapi.py:1059]: PTAL

**Comment by erlend-aasland** [comment]
Putting this on hold until issue 42213 PR's are merged.

**Comment by erlend-aasland** [comment]
Marking this as draft again; I'd like to reverse the responsibilities of the two processes in the test case. The current test case makes it hard to measure coverage.

**Comment by erlend-aasland** [comment]
FYI, @vstinner / @pablogsal: I've rewritten the test so that the code that exercises the target code path executes in the main process; the supporting code now executes in the child process. This makes it easier to measure code coverage. (There are workarounds, but this is the easiest solution, at least for `llvm-cov`.)

PTAL :)

**Comment by pablogsal** [comment]
I'm restarting the ASAN test

**Comment by pablogsal** [comment]
As I am on vacation I, unfortunately, don't have the full bandwidth to review for the next 2 weeks. Maybe @vstinner can take a look. Otherwise, feel free to ping me in 2 weeks :)

**Comment by erlend-aasland** [comment]
@pablogsal Pinging after two weeks :)

**Review by pablogsal — APPROVED** [review]
LGTM

Great job! 🎉

**Comment by pablogsal** [comment]
@erlend-aasland Does this need backports? If not, please close the issue :)

**Comment by erlend-aasland** [comment]
It's a bugfix, so I'd backport to 3.10 and 3.9.

**Comment by erlend-aasland** [comment]
Thank you so much for reviewing, Pablo and Victor!

**Pull request merged**
